Sprint 2/step 2

Neverhood93 opened the pull request:
Добавлены библиотеки react-redux redux-thunk redux @reduxjs/toolkit react-dnd react-dnd-html5-backend
Созданы хранилища Redux для хранения:
списка всех полученных ингредиентов,
списка всех ингредиентов в текущем конструкторе бургера,
объект текущего просматриваемого ингредиента,
объект созданного заказа.
Настроен dnd-react для перетаскивания ингредиента в конструктор и перемещения ингредиента в рамках конструктора
Добавлены api для получения ингредиентов с сервера и создания заказа.
Скорректирован UI для ингредиентов и конструктора.

Neverhood93 committed:
Добавлены библиотеки react-redux redux-thunk redux @reduxjs/toolkit react-dnd react-dnd-html5-backend.
Добавлен store.
Для получения ингредиентов создан Api, экшены, редьюсер через библиотеку  @reduxjs/toolkit
Получение ингредиентов в App изменено на отправление экшена через диспетчерезацию.

Neverhood93 committed:
Реализована навигация по ингредиентам.

Neverhood93 committed:
Подключено dnd к App.

Neverhood93 committed:
Добавлено хранилище для конструктора.

Neverhood93 committed:
Подготовлен черновик для реализации dnd перетаскивания ингредиентов в конструктор.

Neverhood93 committed:
Добавлено ограничение, чтобы без булочки нельзя было собирать бургер.

Neverhood93 committed:
Добавлена возможность удаления ингредиента из конструктора.

Neverhood93 committed:
Добавлен счетчик выбранных ингредиентов.

Neverhood93 committed:
Черновик для функционала перемещения ингредиента в рамках списка в конструкторе.

Neverhood93 committed:
Рефакторинг BurgerConstructorUnlockList и DraggableIngredient.

Neverhood93 committed:
Реализован функционал подсчета итоговой стоимости заказа.

Neverhood93 committed:
Объект текущего просматриваемого ингредиента перенесен в Redux.

Neverhood93 committed:
Добавлена реализация создания заказа через Api.

Neverhood93 committed:
Добавлена очистка заказа при ошибке от сервера.

mikebars1995 commented on src/services/ingredients/actions.tsx:
```diff
@@ -0,0 +1,15 @@
+>("ingredients/loadIngredients", async (_, { rejectWithValue }) => {
+  try {
+    return await getIngredients();
+  } catch (error: any) {
```
**Нужно исправить**


В `createAsyncThunk` не нужно указывать блок `catch`, так как `createAsyncThunk` сам ловит ошибки и отображает их в поле `rejected`. А его уже обрабатывать можно в `createSlice`

mikebars1995 commented on src/utils/ingredients-api.tsx:
```diff
@@ -0,0 +1,27 @@
+import { BurgerIngredient } from "../types/types";
+
+const ingredientsApiConfig = {
+  baseUrl: "https://norma.nomoreparties.space/api/ingredients",
```
**Нужно исправить**

`'https://norma.nomoreparties.space/api` повторяется . Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

mikebars1995 commented on src/utils/ingredients-api.tsx:
```diff
@@ -0,0 +1,27 @@
+  data: BurgerIngredient[];
+}
+
+const getResponse = async (res: Response): Promise<BurgerIngredient[]> => {
```
**Нужно исправить**


Нужно сделать универсальную функцию проверки ответа от сервера, чтобы не дублировать код

mikebars1995 commented on src/utils/orders-api.tsx:
```diff
@@ -0,0 +1,27 @@
+  },
+};
+
+const getResponse = async (res: Response): Promise<OrderResponse> => {
```
**Нужно исправить**


Нужно сделать универсальную функцию проверки ответа от сервера, чтобы не дублировать код

mikebars1995 commented on src/services/burger-constructor/reducer.tsx:
```diff
@@ -0,0 +1,66 @@
+        return {
+          payload: {
+            ...ingredient,
+            uniqueId: uuidv4(),
```
**Отлично**

Отлично, что указываете **уникальный** `uuid` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`

mikebars1995 reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на *гитхабе* все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная


Что сделано отлично:


* Отлично, что указываете **уникальный** `uuid` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`
* Отличная логика `redux` и `DnD`

Что нужно исправить:

* В `createAsyncThunk` не нужно указывать блок `catch`, так как `createAsyncThunk` сам ловит ошибки и отображает их в поле `rejected`. А его уже обрабатывать можно в `createSlice`
* `'https://norma.nomoreparties.space/api` повторяется . Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.
* Нужно сделать универсальную функцию проверки ответа от сервера, чтобы не дублировать код
* В ` BurgerIngredients` и `BurgerConstructor` больше не нужно передавать в `props` массив ингредиентов. Для этого придумали как раз **redux**. Нужно теперь все брать из редакса внутри компонентов. И теперь нужно удалить **типизацию** массива ингредиентов в `BurgerIngredients` и `BurgerConstructor`



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

**Комментарии `Можно лучше` не обязательны к исправлению прямо сейчас, это рекомендации**

Удачного рефакторинга кода.

Neverhood93 committed:
Исправлены замечания после ревью: скорректированы api, редьюсеры, экшены.

Neverhood93 committed:
Исправлены замечания после ревью:
В  BurgerIngredients и BurgerConstructor больше не нужно передавать в props массив ингредиентов. Для этого придумали как раз redux. Нужно теперь все брать из редакса внутри компонентов. И теперь нужно удалить типизацию массива ингредиентов в BurgerIngredients и BurgerConstructor

mikebars1995 reviewed (commented):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
